File: Creational/AbstractFactory/Python/Example/ui_factory.py
from abc import ABC, abstractmethod


class UIFactory(ABC):
    """
    Abstract Factory for creating UI components.
    Each concrete factory creates a family of UI components with a specific theme.
    """
    
    @abstractmethod
    def create_button(self):
        """Create a button component"""
        pass
    
    @abstractmethod
    def create_dialog(self):
        """Create a dialog component"""
        pass
    
    @abstractmethod
    def create_text_field(self):
        """Create a text field component"""
        pass

File: Creational/AbstractFactory/TypeScript/Example/UIFactory.ts
import { Button } from "./Button";
import { Checkbox } from "./Checkbox";

/**
 * UIFactory - Abstract base class for all UI factories
 * Defines the interface for creating families of UI components
 */
export abstract class UIFactory {
  /**
   * Creates a button component
   */
  abstract createButton(text: string): Button;

  /**
   * Creates a checkbox component
   */
  abstract createCheckbox(label: string): Checkbox;

  /**
   * Gets the design system name this factory creates
   */
  abstract getDesignSystem(): string;

  /**
   * Gets the theme name
   */
  abstract getTheme(): string;
}
